### DISCUSSION TITLE
Alignment between Dagger and Cue roadmaps

### DISCUSSION THREAD

**Discussion by shykes** [dagger/dagger#857] [General]
There is much overlap between the roadmaps of Dagger and Cue. I thought it would be useful to have a catalog of those issues, and perhaps a centralized discussion thread would help us assemble that catalog together.

I’m not 100% sure whether this should be an issue or a discussion… But since it’s quite open ended, and it’s not obvious when this discussion should evet be “closed”… I am trying the discussion format.

Below is a list of tasks or workstreams which are needed by both Cue and Dagger. Please suggest more, or start discussions on existing items as needed!

## 1. Cue modules

This one is obvious… Dagger can be programmed with Cue. The ability to share and reuse software components as Cue packages is crucial to the Dagger developer experience.

## 2. Package documentation

 The ability to generate great documentation for any Cue package, similar to `go doc`, would be very useful. Dagger has implemented a fill-in with `dagger doc`.

## 3. Information on data flow

This is a sub-item of package documentation. `dagger doc` relies on a concept of “inputs” and “outputs” which are specific to Dagger… but I believe all Cue developers would benefit from knowing which fields of a given definition are required to be set, and which fields are derived from input fields. This is equivalent to distinguishing parameters from return values in a regular programming language. Developers need help understanding the flow of data, especially in code written by someone else.

## 4. Embedding

Dagger developers in particular, and all Cue developers in general, will benefit from seamlessly combining the contents of Cue and non-Cue files bundled together in the same Cue package. In the case of Dagger, this facilitates packaging hybrid components for the Dagger runtime: imperative code (shell, Go, Pythong, Javascript..) executed in a declaratively specified pipeine (Cue).

## 5. Performance

Some Dagger configurations are very slow to load, both with `dagger` and `cue`. Some of these performance issues have been solved in past Cue updates; others are already known; more will be discovered in the future as Dagger adoption grows.

## 6. Editor integration

A good developer experience requires good integration with code editors. That means implementing a language server, and integrating with as many editors as possible, starting with the most popular ones. The goal is that seamless integration with existing editors is not only technically possible, but easy to find and setup - or even better, just works out of the box.

## 7. Schema migrations

Cue encourages sharing and reusing definitions. Inevitably, some of these definitions will change in a backwards-incompatible way, causing their downstream consumers to break. In theory, Cue has the capability to help solve this problem. For example, the`cue fix` will magically modify your configuration to adapt to even major language changes. Perhaps the same could be possible one day for package changes?

## 8. Hidden fields API

Dagger encourages the use of hidden fields (`_foo`) for information private to a package. This helps keep configurations clean and readable. However, an important feature of Dagger is that it can fill “blank” fields dynamically by executing external code. At the moment Dagger can fill regular fields in this way, but *not* hidden fields, because the Go Cue API does not allow it.

The result is that Dagger developers cannot must choose between A) making a field private, and B) filling it dynamically. This makes Dagger configurations less clean and readable than they could be.

## 9. Catch broken references to outside packages

This is a very specific issue, but it is a constant source of friction in the Cue development experience; and therefore the Dagger development experience.

In short: sometimes Cue catches broken references (references to values that don’t exist) and that is very useful; but sometimes Cue does not catch them, and that causes a very frustrating experience. A better experience would be for all broken references to be caught by `cue eval`, so that the developer can fix them.

This has been reported upstream, and I believe is a *WONTFIX*. That is a mistake.

Here is a snippet:

```
import “strings” // could be any package, external or built-in

broken1: #Nope // Broken reference, causes an error (good experience)
broken2: strings.#Nope // Broken reference, no error (bad experience!)
```

## 10. Stacking default values

CUE does not support stacking default values: setting multiple defaut values for the same field, in different scopes. This is discussed in a follow-up thread in #1243. This is by design, because it would break a cardinal rule of CUE whereby *order doesn’t matter*.

As justified as it is, this design decision can lead to developer pain. In practice it means that authors of reusable packages cannot use default values without causing their abstraction to leak. The pain can be managed, but we should monitor it, in case a solution or mitigation opportunity emerges.

cc @aluzzardi @samalba @myitcv @mpvl @jlongtine @sdboyer

**Comment by myitcv** [comment]
One item we might considering adding to this list is consolidating a plan for CUE documentation, in particular for newcomers to the language. 

Brainstorming what might fall under that umbrella (definitely not exhaustive):

* Tour of CUE (like the [Tour of Go](https://tour.golang.org/welcome/1))
* The [CUE playground](https://cuelang.org/play) needs a refresh - [cue-lang/cue#1150](https://github.com/cue-lang/cue/issues/1150) and friends
* Tutorials
  * Language-based tutorials as part of the Tour of CUE
  * [CUEtorials](https://cuetorials.com/), or a variation on that theme
  * Play with CUE for learning the tooling aspect in an interactive way (c.f. [Play with Go](https://play-with-go.dev/))
  * (Interactive) real-life examples using:
    * `cmd/cue`
    * `cue cmd` (`cuerun`)
    * The Go API
* [Effective CUE](https://github.com/cue-lang/cue/issues/1013) (c.f. [Effective Go](https://golang.org/doc/effective_go))
* ... 

Improving the core CUE documentation (independent of the Dagger docs) has the obvious benefit of helping to grow/support the wider CUE ecosystem which has derivative benefits for Dagger, but also means we can standardise on an approach with respect to:

* translations in other languages
* format, styling
* automated testing (across CUE versions, Go versions etc)
* ...

**Comment by shykes** [comment]
I added a new point: *8. Hidden fields API*.

**Reply by myitcv** [reply]
@shykes as far as I'm aware things should be working with hidden fields. e.g.

```
go mod tidy
go run main.go
cmp stdout stdout.golden

-- cue.mod/module.cue --
module: "mod.com"
-- go.mod --
module mod.com

go 1.16

require cuelang.org/go v0.4.0
-- main.go --
package main

import (
	"fmt"

	"cuelang.org/go/cue"
	"cuelang.org/go/cue/cuecontext"
	"cuelang.org/go/cue/load"
)

func main() {
	ctx := cuecontext.New()
	bps := load.Instances([]string{"."}, nil)
	v := ctx.BuildInstance(bps[0])
	x := v.LookupPath(cue.MakePath(cue.Str("x")))
	fmt.Printf("before: %v\n", x)
	x = x.FillPath(cue.MakePath(cue.Hid("_x", "mod.com/p")), 1)
	fmt.Printf("after:  %v\n", x)
}
-- p/p.cue --
package p

#Def: {
	_x: int
	y:  int
	z:  _x + y
}
-- x.cue --
package x

import "mod.com/p"

x: p.#Def & {
	y: 2
}
-- stdout.golden --
before: {
	y: 2
	z: _x + y
}
after:  {
	y: 2
	z: 3
}
```

Please can you raise an issue if you're seeing something that isn't working, or if there are gaps in any parts of the API?

**Reply by shykes** [reply]
I didn’t realize that `cue.Hid` already existed! That looks like it will solve our issue. We’ll confirm asap. Thanks!

cc @aluzzardi @samalba

**Comment by shykes** [comment]
I added a new point: *9. Catch broken references to outside packages*

**Reply by myitcv** [reply]
That's likely covered by https://github.com/cue-lang/cue/issues/629

**Comment by shykes** [comment]
Added point 10: stacking default values. This has been discussed elsewhere, we had just omitted to add it to this “radar”.
